### PR TITLE
storage, k8s: Fallback to kube-system on empty ns

### PR DESCRIPTION
**Pull request by qinqon**
When the kubeconfig has no namespace on the current context the request
to handle leader election resources fail with missing namespace since a
request like "/foobar" cannot be send. This change use "kube-system" as
the fallback one if the namespace is empty.

### CODE REVIEW

File: pkg/storage/kubernetes/ipam.go [hunk 1]
```diff
@@ -35,6 +35,9 @@ func NewKubernetesIPAM(containerID string, ipamConf whereaboutstypes.IPAMConfig)
 		return nil, err
 	} else if ctx, ok := cfg.Contexts[cfg.CurrentContext]; ok && ctx != nil {
 		namespace = ctx.Namespace
+		if namespace == "" {
+			namespace = "kube-system"
```
Comment (maiqueb) [inline pkg/storage/kubernetes/ipam.go:39]: I think you should use the `metav1.NamespaceSystem` constant instead of this hard-coded string. 

I would say I also miss a unit test for this, but the whole pkg as none .... Not good.

**Review by maiqueb — CHANGES_REQUESTED** [review]
Having a fall-back / default namespace is a good call.

Never the less, I think you should use a constant instead of an hard-coded string.

Furthermore, I have a question: how can the kubeconfig have no namespace on the current context ? I am 100% sure I have been there, but I don't remember right now how I reached that situation.

**Review by maiqueb — APPROVED** [review]
Thanks.

File: pkg/storage/kubernetes/ipam.go [hunk 2]
```diff
@@ -35,6 +35,9 @@ func NewKubernetesIPAM(containerID string, ipamConf whereaboutstypes.IPAMConfig)
 		return nil, err
 	} else if ctx, ok := cfg.Contexts[cfg.CurrentContext]; ok && ctx != nil {
 		namespace = ctx.Namespace
+		if namespace == "" {
+			namespace = metav1.NamespaceSystem
+		}
```
Comment (maiqueb) [inline pkg/storage/kubernetes/ipam.go:40]: actually , would you mind exporting this into a function ? 

Something like

```golang
func wbNamespace(ctx *clientcmdapi.Context) string {
	namespace := ctx.Namespace
	if namespace == "" {
		return metav1.NamespaceSystem
	}
	return namespace
}
```

Comment (maiqueb) [reply pkg/storage/kubernetes/ipam.go]: I think that's wrong. There are 2 deployment paths for whereabouts: 
- a simple CNI plugin - which relies on having a kubeconfig path - i.e. it doesn't have an in cluster config
- a controller / cronjob running in a pod, which rely on the in cluster config.

The above are exclusive: the developer should be explicit and know which to call. By doing this change, you are imo mixing these two types of clients, thus making it more confusing, and hard to reason about.

File: pkg/storage/kubernetes/ipam.go [hunk 3]
```diff
@@ -31,17 +32,27 @@ import (
...
-	} else if ctx, ok := cfg.Contexts[cfg.CurrentContext]; ok && ctx != nil {
-		namespace = ctx.Namespace
+	var kubernetesClient *Client
+	if ipamConf.Kubernetes.KubeConfigPath == "" {
+		var err error
+		kubernetesClient, err = NewClient(storage.RequestTimeout)
+		if err != nil {
+			return nil, fmt.Errorf("failed instantiating kubernetes client: %v", err)
+		}
+		namespace = metav1.NamespaceSystem
```
Comment (maiqueb) [inline pkg/storage/kubernetes/ipam.go:42]: this is wrong imo. See https://github.com/k8snetworkplumbingwg/whereabouts/pull/221#discussion_r870483371 .

**Review by maiqueb — CHANGES_REQUESTED** [review]

File: pkg/storage/kubernetes/ipam.go [hunk 4]
```diff
@@ -31,15 +32,16 @@ import (
...
 	} else if ctx, ok := cfg.Contexts[cfg.CurrentContext]; ok && ctx != nil {
-		namespace = ctx.Namespace
+		namespace = wbNamespaceFromCtx(ctx)
 	} else {
 		return nil, fmt.Errorf("k8s config: namespace not present in context")
 	}
-
-	kubernetesClient, err := NewClientViaKubeconfig(ipamConf.Kubernetes.KubeConfigPath, storage.RequestTimeout)
+	var err error
+	kubernetesClient, err = NewClientViaKubeconfig(ipamConf.Kubernetes.KubeConfigPath, storage.RequestTimeout)
```
Comment (maiqueb) [inline pkg/storage/kubernetes/ipam.go:44]: Why is this better - and even related to this PR ?

File: pkg/storage/kubernetes/ipam.go [hunk 5]
```diff
@@ -19,6 +19,7 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
```
Comment (maiqueb) [inline pkg/storage/kubernetes/ipam.go:22]: why is this needed ?

**Review by maiqueb — COMMENTED** [review]
One quick question: I think this might be a leftover.

File: pkg/storage/kubernetes/ipam.go (hunk 5 above)
Comment (maiqueb) [reply pkg/storage/kubernetes/ipam.go:22]: oh. Just noticed https://github.com/k8snetworkplumbingwg/whereabouts/pull/221/files#diff-7a77fc45b536e504372eba572994f95f478aa250d4cdb936bdc3d1018c86b09eR548 

My bad.

**Review by maiqueb — APPROVED** [review]
tks

**Review by dougbtv — APPROVED** [review]
